### PR TITLE
Implement cabal new-build --only-dependencies

### PR DESCRIPTION
**Pull request by dcoutts**
The flag was already there, this just hooks it up with the logic.

It works in the obvious way given how new-build works. That is, it does not affect solving, it's just a matter of changing what bits we chose to build, much like the logic for selecting which bits of the install plan to execute based on the targets to build.

The difference from pruning for the selected targets (which is just taking a dep closure) is that for `--only-dependencies` it does not always make sense so it can fail. The logic for this already existed in the old `cabal install` code path, so this just ports it over. It's a little different since it works on the `ElaboratedInstallPlan` rather than the `SolverInstallPlan`.

If it fails it gives us something like:

    $ cabal new-build --dry network ./hackage-repo-tool/ --dry --only-dependencies
    Cannot select only the dependencies (as requested by the '--only-dependencies'
    flag), the package network-2.6.2.1 is required by a dependency of one of the
    other targets.

I've only lightly tested this but it appears to work. I'm relying on @hvr to test it properly. If anyone would like to contribute some unit tests that'd be nice :-). We'd want one that checks fewer packages are built, and one that checks that the corner case where it cannot work is handled properly (ie throws an exception).

### CODE REVIEW

File: cabal-install/Distribution/Client/CmdRepl.hs [hunk 1]
```diff
@@ -73,7 +73,25 @@ replAction (configFlags, configExFlags, installFlags, haddockFlags)
...
+                 ++ "You may wish to use 'build --only-dependencies' and then "
+                 ++ "use 'repl'."
+            -- Interpret the targets on the command line as repl targets
+            -- (as opposed to say build or haddock targets).
+            selectTargets
+              verbosity
+              ReplDefaultComponent
+              ReplSpecificComponent
+              userTargets
+              False -- onlyDepencencies, always False for repl
```
Comment (ezyang) [inline cabal-install/Distribution/Client/CmdRepl.hs]: onlyDependencies

Comment (dcoutts) [inline cabal-install/Distribution/Client/CmdRepl.hs]: Doh! I should turn on the spellcheck by default :-)

File: cabal-install/Distribution/Client/ProjectPlanning.hs [hunk 2]
```diff
@@ -1997,6 +2000,86 @@ componentOptionalStanza (Cabal.CTestName  _) = Just TestStanzas
...
+-- | Try to remove the given targets from the install plan.
+--
+-- This is not always possible.
+--
+pruneInstallPlanToDependencies :: Set UnitId
+                               -> ElaboratedInstallPlan
+                               -> Either CannotPruneDependencies
+                                         ElaboratedInstallPlan
+pruneInstallPlanToDependencies pkgTargets installPlan =
+    assert (all (isJust . InstallPlan.lookup installPlan)
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: BTW, we should add `-fno-ignore-asserts` to `cabal.project`.

Comment (dcoutts) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: Yeah. I added it in the integration-tests2 previously, but we should have it in integration-tests and unit-tests

File: cabal-install/Distribution/Client/ProjectOrchestration.hs [hunk 3]
```diff
@@ -268,10 +269,11 @@ runProjectBuildPhase verbosity ProjectBuildContext {..} =
 selectTargets :: Verbosity -> PackageTarget
               -> (ComponentTarget -> PackageTarget)
               -> [UserBuildTarget]
+              -> Bool
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectOrchestration.hs]: I'd use a special-purpose `OnlyDependencies` ADT instead of a Bool, but that's a very minor thing.

**Comment by 23Skidoo** [comment]
LGTM.

**Comment by hvr** [comment]
@dcoutts This seems to work nice, except for the `cabal new-build` case when using a `cabal.project` file (and no `.cabal` file in the top-level folder & with CWD=top-level-folder) and thus causing it to implicitly build all packages: in this case `cabal new-build --deps` seems to consider nothing to be a dependency goal, e.g.

```
$ cd /tmp/foo
$ cabal get uhttpc-0.1.1.0
Unpacking to uhttpc-0.1.1.0/
$ echo "packages: uhttpc-0.1.1.0"  > cabal.project
$ cabal new-build --dry
Resolving dependencies...
In order, the following would be built (use -v for more details):
 - bytestring-lexing-0.5.0.2 (lib:bytestring-lexing, lib:bytestring-lexing) (requires build)
 - uhttpc-0.1.1.0 (lib) (first run)
 - uhttpc-0.1.1.0 (exe:uhttpc-bench) (first run)
$ cabal new-build --dep
Up to date
```

btw, is `cabal new-build uhttpc-0.1.1.0/` supposed to work?

```
$ pwd
/tmp/foo

$ cabal new-build uhttpc-0.1.1.0/ 
cabal: Unknown build target 'uhttpc-0.1.1.0/'.
The project has no package directory 'uhttpc-0.1.1.0/'.
Perhaps you meant the package directory ../foo/uhttpc-0.1.1.0 or
../foo/uhttpc-0.1.1.0
```

**Comment by ezyang** [comment]
@hvr Do you consider that bug a blocker to merging this?

**Comment by hvr** [comment]
@ezyang not at all! I'm all for merging this, as it's strictly an improvement as-is

**Comment by ezyang** [comment]
OK merged. @hvr said he will write a follow up ticket.
